ArC - introduce ObserverTransformer build extension

mkouba opened the pull request.

mkouba committed:
ArC - introduce ObserverTransformer build extension

manovotn commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationsTransformationContext.java:
```diff
@@ -0,0 +1,62 @@
+        this.annotations = annotations;
+    }
+
+    public Collection<AnnotationInstance> getAllAnnotations() {
```
I'd add similar javadoc to the one we have in `InjectionPointsTransformer` for `TransformationContext#getAllAnnotations`.

mkouba replied:
Actually, we have the javadoc on the user-facing interfaces such as `io.quarkus.arc.processor.ObserverTransformer.TransformationContext`.

manovotn reviewed (commented):
The PR looks good, added some minor comments.
We'll also need to document this.
Last but not least, +1 for the refactoring.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Transformation.java:
```diff
@@ -1,116 +1,20 @@
- * Convenient helper class.
- */
-public final class Transformation {
+public final class Transformation extends AbstractAnnotationsTransformation<Transformation, Collection<AnnotationInstance>> {
```
Yes, that's the exact reason...

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverTransformer.java:
```diff
@@ -0,0 +1,121 @@
+ * Allows a build-time extension to:
+ * <ul>
+ * <li>override the set of qualifiers and priority of any observer method,</li>
+ * <li>instruct the container to ignore the observer method.</li>
```
Ok.

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverTransformer.java:
```diff
@@ -0,0 +1,121 @@
+
+    }
+
+    interface ObserverTransformation extends AnnotationsTransformation<ObserverTransformation> {
```
Hm, I'm not so sure. Async observers are invoked in a different context. But the truth is that it's possible to change the transaction phase which could eventually (transactional observers are not implemented yet) result in an async invocation as well.

manovotn replied:
> Async observers are invoked in a different context. 

What do you mean by different context?

I just thought that since we allow to change reception and tx phases, this could work as well.
It shouldn't change/break the logic inside the observer either.

mkouba replied:
> What do you mean by different context?

I mean different thread, transaction context, CDI contexts...

> we allow to change reception and tx phases

Reception is ok - does not change the context. Transaction phase - yep, that's what I mentioned in the previous comment.

mkouba committed:
Docs - observer transformation

mkouba committed:
SmallRye FT - initialize Hystrix during RUNTIME_INIT

- veto HystrixInitializer.init() observer
- add HystrixInitializerStartup
- resolves #6024

manovotn commented on extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/HystrixInitializerStarter.java:
```diff
@@ -0,0 +1,23 @@
+     * @param initializer
+     */
+    void startup(@Observes StartupEvent event, HystrixInitializer initializer) {
+        initializer.toString();
```
Might be worth adding a note that we need to trigger any bean-specific method or `toString` in order to enforce bean creation. While it's obvious for us, I can see a lot of non-CDI folks staring at this suspiciously ;-)

mkouba replied:
Ok, I can add a link to the "Eager Instantiation of Beans" section.

mkouba replied:
Eh, sorry not a link but a comment.

gwenneg commented:
Thanks for this enhancement @mkouba!
